benchmarks/persistence/persistence.py

```diff
@@ -0,0 +1,205 @@
+"""Classes for persisting benchmark results."""
+
+import json
+import os
+from datetime import datetime
+from enum import Enum
+from pathlib import Path
+from typing import Protocol
+
+import boto3
+import boto3.session
+from attr import define, field
+from attrs.validators import instance_of
+from git import Repo
+
+from benchmarks import Benchmark, Result
+
+
+class PathStrategy(Enum):
+    """The way a path extension is constructed."""
+
+    Hierarchical = 1
+    """Refers to the hierarchical path construction of folders."""
+
+    Flat = 2
+    """Refers to the flat path construction of a single file name."""
+
+
+@define
+class PathConstructor:
+    """A class to construct the hierarchical path of a result object."""
+
+    benchmark_name: str = field(validator=instance_of(str))
+    """The name of the benchmark for which the path should be constructed."""
+
+    latest_baybe_tag: str = field(validator=instance_of(str))
+    """The latest BayBE version tag"""
+
+    commit_hash: str = field(validator=instance_of(str))
+    """The latest commit hash of the current checkout branch."""
+
+    execution_date_time: datetime = field(validator=instance_of(datetime))
+    """The date and time when the benchmark was executed."""
+
+    branch: str = field(validator=instance_of(str), init=False)
+    """The branch currently checked out."""
+
+    @branch.default
+    def _default_branch(self) -> str:
+        repo = Repo(search_parent_directories=True)
+        current_branch = repo.active_branch.name
+        return current_branch
+
+    def _string_sanitizer(self, string: str) -> str:
+        """Replace not allowed chars for filesystems in a string."""
+        ALLOWED_CHARS = (
+            "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-."
+        )
+        return "".join([char if char in ALLOWED_CHARS else "-" for char in string])
+
+    def get_path(self, strategy: PathStrategy) -> str:
+        """Construct the path of a result object.
+
+        Create the path depending on the chosen strategy.
+        It will make use of the experiment identifier,
```

Comment:
Replace future tense by present tense since you are describing what this __does__.